### PR TITLE
[Docs] Translate dataset tutorial

### PR DESCRIPTION
**Pull request by C1rN09**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: docs/en/tutorials/dataset.md [hunk 1]
```diff
@@ -0,0 +1,200 @@
+# Dataset and DataLoader
+
+```{hint}
+If you have never been exposed to PyTorch's dataset and dataloader, you are recommended to read through [PyTorch official tutorial](https://pytorch.org/tutorials/beginner/basics/data_tutorial.html) to get familiar with some basic concepts.
+```
+
+Datasets and DataLoaders are necessary components in MMEngine's training pipeline. They are conceptually derived from and consistent with PyTorch. Typically, a dataset defines the quantity, parsing and pre-processing of the data, while a dataloader iteratively load data according to settings such as `batch_size`, `shuffle`, `num_workers`, etc. Datasets are encapsuled with dataloaders and they together constitute the data source.
```
Comment (HAOCHENYE) [inline docs/en/tutorials/dataset.md:7]:
```suggestion
Datasets and DataLoaders are necessary components in MMEngine's training pipeline. They are conceptually derived from and consistent with PyTorch. Typically, a dataset defines the quantity, parsing and pre-processing of the data, while a dataloader iteratively load data according to settings such as `batch_size`, `shuffle`, `num_workers`, etc. Datasets are encapsulated with dataloaders and they together constitute the data source.
```

File: docs/en/tutorials/dataset.md [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+In this tutorial, we will step through their usage in MMEngine runner from outside (dataloader) to inside (dataset) and give some practical examples. After reading through this tutorial, you will be able to:
```
Comment (HAOCHENYE) [inline docs/en/tutorials/dataset.md:9]:
```suggestion
In this tutorial, we will step through their usage in MMEngine runner from outside (dataloader) to the inside (dataset) and give some practical examples. After reading through this tutorial, you will be able to:
```

File: docs/en/tutorials/dataset.md [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+- Master the configuration of dataloaders in MMEngine
+- Learn to use existing datasets (e.g. those from `torchvision`) from config files
+- Know about building and using your own dataset
+
+## Details on dataloader
+
+Dataloaders can be configured in MMEngine's `Runner` with 3 arguments:
```
Comment (HAOCHENYE) [inline docs/en/tutorials/dataset.md:17]: dataloader, Dataloader -> DataLoader

File: docs/en/tutorials/dataset.md [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+- `train_dataloader`: Used in `Runner.train()` to provide training data for models
+- `val_dataloader`: Used in `Runner.val()` or in `Runner.train()` at regular intervals for model evaluation
+- `test_dataloader`: Used in `Runner.test()` for the final test
+
+MMEngine has full support for PyTorch native `DataLoader` objects. Therefore, you can simply pass your valid, already built dataloaders to the runner, as shown in [getting started in 15 minutes](../get_started/15_minutes.md). Meanwhile, thanks to the [Registry Mechanism](../advanced_tutorials/registry.md) of MMEngine, those arguments also accepts `dict`s as inputs, as illustrated in the following example (referred to as example 1). The keys in the dictionary correspond to arguments in DataLoader's init function.
```
Comment (HAOCHENYE) [inline docs/en/tutorials/dataset.md:23]:
```suggestion
MMEngine has full support for PyTorch native `DataLoader` objects. Therefore, you can simply pass your valid, already built dataloaders to the runner, as shown in [getting started in 15 minutes](../get_started/15_minutes.md). Meanwhile, thanks to the [Registry Mechanism](../advanced_tutorials/registry.md) of MMEngine, those arguments also accept `dict`s as inputs, as illustrated in the following example (referred to as example 1). The keys in the dictionary correspond to arguments in DataLoader's init function.
```

File: docs/en/tutorials/dataset.md [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+```{note}
+For more configurable arguments of the `DataLoader`, please refer to [PyTorch API documentation](https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader)
+```
+
+```{note}
+If you are interested in the details of the building procedure, you may refer to [build_dataloader](mmengine.runner.Runner.build_dataloader)
+```
+
+You may find example 1 differs from that in [getting started in 15 minutes](../get_started/15_minutes.md) in some arguments. Indeed, due to some obscure conventions in MMEngine, you can't seamlessly switch it to a dict by simply replace `DataLoader` with `dict`. We will discuss the differences between our convention and PyTorch's in the following sections, in case you run into trouble when using config files.
```
Comment (HAOCHENYE) [inline docs/en/tutorials/dataset.md:48]:
```suggestion
You may find example 1 differs from that in [getting started in 15 minutes](../get_started/15_minutes.md) in some arguments. Indeed, due to some obscure conventions in MMEngine, you can't seamlessly switch it to a dict by simply replacing `DataLoader` with `dict`. We will discuss the differences between our convention and PyTorch's in the following sections, in case you run into trouble when using config files.
```

File: docs/en/tutorials/dataset.md [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+
+### sampler and shuffle
+
+One obvious difference is that we add a `sampler` argument to the dict. This is because we **require `sampler` be explicitly specified** when using dict as dataloader. Meanwhile, `shuflle` is also removed from `DataLoader` arguments, because it conflicts with `sampler` in PyTorch, as referred to [PyTorch DataLoader API docs](https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader).
```
Comment (HAOCHENYE) [inline docs/en/tutorials/dataset.md:52]:
```suggestion
One obvious difference is that we add a `sampler` argument to the dict. This is because we **require `sampler` to be explicitly specified** when using dict as dataloader. Meanwhile, `shuflle` is also removed from `DataLoader` arguments, because it conflicts with `sampler` in PyTorch, as referred to [PyTorch DataLoader API docs](https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader).
```

File: docs/en/tutorials/dataset.md [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+        batch_size=32,
+        sampler=sampler,
+        dataset=dataset,
+        collate_fn=default_collate
+    )
+)
+```
+
+```{warning}
+The equivalence of the above codes holds only if: 1) you are training with a single process, and 2) no `randomness` argument is passed to the runner. This is due to the fact that `sampler` should be built after distributed environment setup to be correct. The runner will guarantee the correct order and proper random seed by applying lazy initialization techniques, which is only possible for dict inputs. Instead, when building sampler manually, it requires extra work and is highly error-prone. Therefore, the code block above is just for illustration and definitely not recommended. We **strongly suggest passing `sampler` as a `dict`** to avoid potential problems.
```
Comment (HAOCHENYE) [inline docs/en/tutorials/dataset.md:77]:
```suggestion
The equivalence of the above codes holds only if: 1) you are training with a single process, and 2) no `randomness` argument is passed to the runner. This is due to the fact that `sampler` should be built after distributed environment setup to be correct. The runner will guarantee the correct order and proper random seed by applying lazy initialization techniques, which is only possible for dict inputs. Instead, when building a sampler manually, it requires extra work and is highly error-prone. Therefore, the code block above is just for illustration and definitely not recommended. We **strongly suggest passing `sampler` as a `dict`** to avoid potential problems.
```

File: docs/en/tutorials/dataset.md [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+```
+
+### Customize your dataset
+
+You are free to customize your own datasets, as you would with PyTorch. You can also copy existing datasets from your previous PyTorch projects. If you want to learn how to customize your dataset, please refer to [PyTorch official tutorials](https://pytorch.org/tutorials/beginner/basics/data_tutorial.html#creating-a-custom-dataset-for-your-files)
+
+### Use MMEngine BaseDataset
+
+Apart from directly using PyTorch native `Dataset` class, you can also use MMEngine's built-in class `BaseDataset` to customize your own one, as referred to [BaseDataset tutorial](../advanced_tutorials/basedataset.md). It make some conventions on the format of annotation files, which makes the data interface more unified and multi-task training more convenient. Meanwhile, `BaseDataset` can easily cooperate with built-in [data transforms](../advanced_tutorials/data_element.md) in MMEngine, which release you from writing one from scratch.
```
Comment (HAOCHENYE) [inline docs/en/tutorials/dataset.md:198]:
```suggestion
Apart from directly using PyTorch native `Dataset` class, you can also use MMEngine's built-in class `BaseDataset` to customize your own one, as referred to [BaseDataset tutorial](../advanced_tutorials/basedataset.md). It makes some conventions on the format of annotation files, which makes the data interface more unified and multi-task training more convenient. Meanwhile, `BaseDataset` can easily cooperate with built-in [data transforms](../advanced_tutorials/data_element.md) in MMEngine, which releases you from writing one from scratch.
```

File: docs/en/tutorials/dataset.md [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+                dict(type='RandomCrop', size=32, padding=4),
+                dict(type='RandomHorizontalFlip'),
+                dict(type='ToTensor'),
+                dict(type='Normalize', **norm_cfg)])
+    )
+)
+```
+
+```{note}
+The above example makes extensive use of the registry mechanism and borrows the [Compose](mmengine.dataset.Compose) module from MMEngine. If you urge to use torchvision dataset in your config files, you can refer to it and make some slight modifications. However, we recommend you to borrow datasets from downstream repos such as [MMDet](https://github.com/open-mmlab/mmdetection), [MMCls](https://github.com/open-mmlab/mmclassification), etc. This may give you better experience.
```
Comment (HAOCHENYE) [inline docs/en/tutorials/dataset.md:189]:
```suggestion
The above example makes extensive use of the registry mechanism and borrows the [Compose](mmengine.dataset.Compose) module from MMEngine. If you urge to use torchvision dataset in your config files, you can refer to it and make some slight modifications. However, we recommend you borrow datasets from downstream repos such as [MMDet](https://github.com/open-mmlab/mmdetection), [MMCls](https://github.com/open-mmlab/mmclassification), etc. This may give you a better experience.
```

File: docs/en/tutorials/dataset.md [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+
+## Details on dataset
+
+Typically, datasets define the quantity, parsing and pre-processing of the data. It is encapsuled in dataloader, allowing the latter to load data in batches. Since we fully support PyTorch `DataLoader`, the dataset is also compatible. Meanwhile, thanks to the registry mechanism, when a dataloader is given as a dict, its `dataset` argument can also be given as a dict, which enables lazy initialization in the runner. This mechanism allows for writing config files.
+
+### Use torchvision datasets
+
+`torchvision` provides various open datasets. They can be directly used in MMEngine as shown in [getting started in 15 minutes](../get_started/15_minutes.md), where a `CIFAR10` dataset is used together with torchvision's built-in data transforms.
+
+However, if you want to use the dataset in config files, registration is needed. What's more, if you also requires data transforms in torchvision, some more registrations are required. The following example illustrates how to do it.
```
Comment (HAOCHENYE) [inline docs/en/tutorials/dataset.md:145]:
```suggestion
However, if you want to use the dataset in config files, registration is needed. What's more, if you also require data transforms in torchvision, some more registrations are required. The following example illustrates how to do it.
```

File: docs/en/tutorials/dataset.md [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+    train_dataloader=dict(
+        ...
+        collate_fn=dict(type='my_collate_func')
+    )
+)
+```
+
+## Details on dataset
+
+Typically, datasets define the quantity, parsing and pre-processing of the data. It is encapsuled in dataloader, allowing the latter to load data in batches. Since we fully support PyTorch `DataLoader`, the dataset is also compatible. Meanwhile, thanks to the registry mechanism, when a dataloader is given as a dict, its `dataset` argument can also be given as a dict, which enables lazy initialization in the runner. This mechanism allows for writing config files.
```
Comment (HAOCHENYE) [inline docs/en/tutorials/dataset.md:139]:
```suggestion
Typically, datasets define the quantity, parsing and pre-processing of the data. It is encapsulated in dataloader, allowing the latter to load data in batches. Since we fully support PyTorch `DataLoader`, the dataset is also compatible. Meanwhile, thanks to the registry mechanism, when a dataloader is given as a dict, its `dataset` argument can also be given as a dict, which enables lazy initialization in the runner. This mechanism allows for writing config files.
```

File: docs/en/tutorials/dataset.md (hunk 3 above)
Comment (C1rN09) [reply docs/en/tutorials/dataset.md:17]: In this tutorial, I try to use Dataset/DataLoader as class names or concepts, and use dataset/dataloader as objects or arguments. Is this reasonable?

**Review by HAOCHENYE — APPROVED** [review]
Approved!!!
image

File: docs/en/tutorials/dataset.md [hunk 12]
```diff
@@ -0,0 +1,200 @@
+# Dataset and DataLoader
+
+```{hint}
+If you have never been exposed to PyTorch's Dataset and DataLoader classes, you are recommended to read through [PyTorch official tutorial](https://pytorch.org/tutorials/beginner/basics/data_tutorial.html) to get familiar with some basic concepts.
+```
+
+Datasets and DataLoaders are necessary components in MMEngine's training pipeline. They are conceptually derived from and consistent with PyTorch. Typically, a dataset defines the quantity, parsing and pre-processing of the data, while a dataloader iteratively load data according to settings such as `batch_size`, `shuffle`, `num_workers`, etc. Datasets are encapsulated with dataloaders and they together constitute the data source.
```
Comment (zhouzaida) [inline docs/en/tutorials/dataset.md:7]:
```suggestion
Datasets and DataLoaders are necessary components in MMEngine's training pipeline. They are conceptually derived from and consistent with PyTorch. Typically, a dataset defines the quantity, parsing, and pre-processing of the data, while a dataloader iteratively loads data according to settings such as `batch_size`, `shuffle`, `num_workers`, etc. Datasets are encapsulated with dataloaders and they together constitute the data source.
```

File: docs/en/tutorials/dataset.md [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+    train_dataloader=dict(
+        ...
+        collate_fn=dict(type='my_collate_func')
+    )
+)
+```
+
+## Details on dataset
+
+Typically, datasets define the quantity, parsing and pre-processing of the data. It is encapsulated in dataloader, allowing the latter to load data in batches. Since we fully support PyTorch `DataLoader`, the dataset is also compatible. Meanwhile, thanks to the registry mechanism, when a dataloader is given as a dict, its `dataset` argument can also be given as a dict, which enables lazy initialization in the runner. This mechanism allows for writing config files.
```
Comment (zhouzaida) [inline docs/en/tutorials/dataset.md:139]:
```suggestion
Typically, datasets define the quantity, parsing, and pre-processing of the data. It is encapsulated in dataloader, allowing the latter to load data in batches. Since we fully support PyTorch `DataLoader`, the dataset is also compatible. Meanwhile, thanks to the registry mechanism, when a dataloader is given as a dict, its `dataset` argument can also be given as a dict, which enables lazy initialization in the runner. This mechanism allows for writing config files.
```
